wagtail/models/reference_index.py

```diff
@@ -0,0 +1,473 @@
+                            value
+                        )
+                    )
+
+        # Extract references from child relations
+        if isinstance(object, ClusterableModel):
+            for child_relation in get_all_child_relations(object):
+                relation_name = child_relation.get_accessor_name()
+                child_objects = getattr(object, relation_name).all()
+
+                for child_object in child_objects:
+                    yield from (
+                        (
+                            to_content_type_id,
+                            to_object_id,
+                            f"{relation_name}.item.{model_path}",
+                            f"{relation_name}.{str(child_object.id)}.{content_path}",
+                        )
+                        for to_content_type_id, to_object_id, model_path, content_path in cls._extract_references_from_object(
+                            child_object
+                        )
+                    )
+
+    @classmethod
+    def _get_content_path_hash(cls, content_path):
+        """
+        Returns a UUID for the given content path. Used to enforce uniqueness.
+
+        Note: MySQL has a limit on the length of fields that are used in unique keys so
+              we need a separate hash field to allow us to support long content paths.
+
+        Args:
+            content_path (str): The content path to get a hash for
+
+        Returns:
+            A UUID instance containing the hash of the given content path
+        """
+        return uuid.uuid5(
+            uuid.UUID("bdc70d8b-e7a2-4c2a-bf43-2a3e3fcbbe86"), content_path
+        )
+
+    @classmethod
+    def create_or_update_for_object(cls, object):
+        """
+        Creates or updates ReferenceIndex records for the given object.
+
+        This method will extract any outbound references from the given object
+        and insert/update them in the database.
+
+        Note: This method must be called within a `django.db.transaction.atomic()` block.
+
+        Args:
+            object (Model): The model instance to create/update ReferenceIndex records for
+        """
+        # Extract new references
+        references = set(cls._extract_references_from_object(object))
+
+        # Find content types for this model and all of its ancestor classes,
+        # ordered from most to least specific
+        content_types = [
+            ContentType.objects.get_for_model(model_or_object, for_concrete_model=False)
+            for model_or_object in ([object] + object._meta.get_parent_list())
+        ]
+        content_type = content_types[0]
+        base_content_type = content_types[-1]
+        known_content_type_ids = [ct.id for ct in content_types]
+
+        # Find existing references in the database so we know what to add/delete
+        existing_references = {
+            (to_content_type_id, to_object_id, model_path, content_path): (
```

Comment:
nitpick from my tired self: this comprehension and the one below (`deleted_reference_ids`) while smart, they also make the code more difficult to parse

Reply:
Added extra comments in 0c2dee315c931a85dd577c088088aa68542ffaf1 and turned the second comprehension into a plain loop - hopefully that makes it easier to follow.